src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs

```diff
@@ -92,37 +93,44 @@ public static void WriteDigitsWithGroupSeparator(ulong value, Span<byte> buffer)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void WriteDigits(uint value, Span<byte> buffer)
+        public static void WriteDigits<TChar>(uint value, Span<TChar> buffer) where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            // We can mutate the 'value' parameter since it's a copy-by-value local.
-            // It'll be used to represent the value left over after each division by 10.
+            Debug.Assert(buffer.Length > 0);
 
             for (int i = buffer.Length - 1; i >= 1; i--)
             {
                 uint temp = '0' + value;
                 value /= 10;
-                buffer[i] = (byte)(temp - (value * 10));
+                buffer[i] = TChar.CreateTruncating(temp - (value * 10));
             }
 
             Debug.Assert(value < 10);
-            buffer[0] = (byte)('0' + value);
+            buffer[0] = TChar.CreateTruncating('0' + value);
         }
 
         /// <summary>
         /// Writes a value [ 0000 .. 9999 ] to the buffer starting at the specified offset.
         /// This method performs best when the starting index is a constant literal.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe void WriteFourDecimalDigits(uint value, Span<byte> buffer, int startingIndex = 0)
+        public static unsafe void WriteFourDecimalDigits<TChar>(uint value, Span<TChar> buffer, int startingIndex = 0) where TChar : unmanaged, IBinaryInteger<TChar>
         {
             Debug.Assert(value <= 9999);
             Debug.Assert(startingIndex <= buffer.Length - 4);
 
             (value, uint remainder) = Math.DivRem(value, 100);
-            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
+            fixed (TChar* bufferPtr = &MemoryMarshal.GetReference(buffer))
             {
-                Number.WriteTwoDigits(bufferPtr + startingIndex, value);
-                Number.WriteTwoDigits(bufferPtr + startingIndex + 2, remainder);
+                if (typeof(TChar) == typeof(char))
+                {
+                    Number.WriteTwoDigits((char*)bufferPtr + startingIndex, value);
+                    Number.WriteTwoDigits((char*)bufferPtr + startingIndex + 2, remainder);
+                }
+                else
+                {
+                    Number.WriteTwoDigits((byte*)bufferPtr + startingIndex, value);
+                    Number.WriteTwoDigits((byte*)bufferPtr + startingIndex + 2, remainder);
+                }
```

Comment:
Why does this need to be split into the `if/else`?

Reply:
We have two overloads of WriteTwoDigits, each of which does a single read/write: the one for bytes reads/writes a ushort and the one for chars reads/writes a uint.  So I split the calling path to use them.  Is there a better solution?

Reply:
Thanks for the explanation!

Pushing the branching/specialization down to `WriteTwoDigits` would help simplify the higher calls a bit more. But I don't have a particular preference.